Nama : Almas Rausan Fikri

In [1]:
import numpy as np
import pandas as pd

# Soal Beginner

## No.1

In [2]:
# Masukan syntax jawaban di sini
data = pd.DataFrame({
    'no': [1,2,3,4,5],
    'nama_siswa': ['Adi', 'Budi', 'Carli', 'Dadan', 'Endang'],
    'umur': [15.0, 16.0, 14.0, None, 16.0],
    'email': ['andi@gmail.com', 'budi@gmail.com', 'carli@rakamin.com', 'dadan@rakamin.com', 'endang@gmail.com'],
    'score_matematika': [80, 90, 100, 95, 97],
    'score_ekonomi': [90.0, 95.0, 100.0, 86.0, None],
    'score_olahraga': [97, 95, 96, 100, 94],
})

data

no nama_siswa  umur              email  score_matematika  score_ekonomi  \
0   1        Adi  15.0     andi@gmail.com                80           90.0   
1   2       Budi  16.0     budi@gmail.com                90           95.0   
2   3      Carli  14.0  carli@rakamin.com               100          100.0   
3   4      Dadan   NaN  dadan@rakamin.com                95           86.0   
4   5     Endang  16.0   endang@gmail.com                97            NaN   

   score_olahraga  
0              97  
1              95  
2              96  
3             100  
4              94

## No.2 

In [3]:
# Masukan sytnax jawaban di sini
data_sorted = data.sort_values('score_matematika', ascending=False) # Terbesar ke terkecil
data_lowest_score = data_sorted.tail(1) # ambil skor terkecil
data_lowest_score

no nama_siswa  umur           email  score_matematika  score_ekonomi  \
0   1        Adi  15.0  andi@gmail.com                80           90.0   

   score_olahraga  
0              97

Siapa nama siswa dengan score_matematika terkecil tersebut? **Adi**

## No.3

In [4]:
# Masukan syntax jawaban di sini
score_ekonomi_nn = data['score_ekonomi'].notnull()
email_gmail = data['email'].str.contains('@gmail.com')

data_siswa_gmail = data[score_ekonomi_nn & email_gmail]
data_siswa_gmail['umur'].mean()

15.5

Berapa umur rata-rata dari siswa di “data_siswa_gmail” tersebut? **15.5**

## No.4

In [5]:
# Masukan syntax jawaban di sini
additional_data = pd.DataFrame({
    'no': [6,7],
    'nama_siswa': ['Fauzan', 'Guntur'],
    'umur': [15, 16],
    'email': ['fauzan@gmail.com', 'guntur@gmail.com'],
    'score_matematika': [80, 90],
    'score_ekonomi': [90, 95],
    'score_olahraga': [97, 95],
})

new_data = data.append(additional_data, ignore_index = True)
new_data['umur'].median()

15.5

berapa median umur mereka sekarang? (setelah append dilakukan) **15.5**

## No.5

In [6]:
# Masukan syntax jawaban di sini
data_address = pd.DataFrame({
    'nama': ['Adi', 'Budi', 'Dadan', 'Endang', 'Guntur', 'Halim'],
    'alamat': ['DKI Jakarta', 'DKI Jakarta', 'Banten', 'Banten', 'Jawa Barat', 'Aceh']
})

df_merged = new_data.merge(data_address, left_on='nama_siswa', right_on='nama')
# How? inner/outer/left/right?

df_merged.groupby(['alamat']).agg({'umur': 'mean'}).sort_values('umur')

umur
alamat           
DKI Jakarta  15.5
Banten       16.0
Jawa Barat   16.0

Pada alamat yang mana, umur rata-rata mereka yang paling kecil? **DKI Jakarta (15.5)**

# Soal Intermediate

## No.6

In [7]:
# Masukan syntax jawaban di sini
airlines = pd.read_excel('data_airlines.xlsx')
airports = pd.read_excel('data_airports.xlsx')
flights = pd.read_excel('data_flights.xlsx')

In [8]:
flights.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      UA           1475      N69819   
1  2015      1    1            4      AA            227      N571AA   
2  2015      1    1            4      AA            357      N3HAAA   
3  2015      1    1            4      MQ           3459      N671MQ   
4  2015      1    1            4      OO           6559      N886AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  ARRIVAL_TIME  \
0            ORD                 MCO                 1600  ...        2146.0   
1            DFW                 MSP                 2040  ...        2321.0   
2            BWI                 MIA                 1726  ...        2003.0   
3            MIA                 PNS                 1635  ...        1751.0   
4            PHX                 PSP                 1943  ...        1946.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          121.0         0          0                  NaN               0.0   
1           26.0         0          0                  NaN               0.0   
2           -9.0         0          0                  NaN               NaN   
3           27.0         0          0                  NaN               6.0   
4           -5.0         0          0                  NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0             0.0           93.0                 28.0            0.0  
1             0.0           26.0                  0.0            0.0  
2             NaN            NaN                  NaN            NaN  
3             0.0           21.0                  0.0            0.0  
4             NaN            NaN                  NaN            NaN  

[5 rows x 31 columns]

In [9]:
flights.duplicated().any()

False

In [10]:
print(flights['YEAR'].unique()) # Cek apakah ada tahun selain 2015
print(flights['CANCELLED'].unique()) # Cek apakah cancelled itu boolean? 1 0, kalau iya, bisa agregasi pakai sum

[2015]
[0 1]


In [11]:
#a. Total penerbangan per bulan selama 2015 (3 point)
#b. Total penerbangan yang CANCELLED per bulan (3 point)
#c. Rata-rata, Minimum dan Maksimum ELAPSED_TIME per bulan (4 point)
soal_6 = flights.groupby('MONTH').agg({ 'FLIGHT_NUMBER': ['count'],
                                        'CANCELLED': ['sum'],
                                        'ELAPSED_TIME': ['mean', 'min', 'max']
                                         }).reset_index()
soal_6.columns = ['2015_MONTH', 'TOTAL_PENERBANGAN', 'TOTAL_CANCELLED', 'ELAPSED_TIME_MEAN', 'ELAPSED_TIME_MIN', 'ELAPSED_TIME_MAX']
soal_6['SUCCESSFUL_FLIGHT'] = soal_6['TOTAL_PENERBANGAN'] - soal_6['TOTAL_CANCELLED']

soal_6[['2015_MONTH', 'TOTAL_PENERBANGAN', 'TOTAL_CANCELLED', 'SUCCESSFUL_FLIGHT', 'ELAPSED_TIME_MEAN', 'ELAPSED_TIME_MIN', 'ELAPSED_TIME_MAX']]

2015_MONTH  TOTAL_PENERBANGAN  TOTAL_CANCELLED  SUCCESSFUL_FLIGHT  \
0            1                798               20                778   
1            2                715               30                685   
2            3                877               22                855   
3            4                824                9                815   
4            5                840                5                835   
5            6                839               18                821   
6            7                937               14                923   
7            8                887                8                879   
8            9                787                1                786   
9           10                856                2                854   
10          11                822                8                814   
11          12                818               17                801   

    ELAPSED_TIME_MEAN  ELAPSED_TIME_MIN  ELAPSED_TIME_MAX  
0          132.153548              31.0             428.0  
1          140.375731              24.0             408.0  
2          142.357562              30.0             622.0  
3          138.420859              31.0             570.0  
4          136.516827              34.0             466.0  
5          139.057317              32.0             398.0  
6          141.110990              30.0             511.0  
7          142.948689              27.0             655.0  
8          137.667516              34.0             495.0  
9          134.366940              31.0             388.0  
10         139.198032              32.0             392.0  
11         144.340426              33.0             453.0

Insight menarik: 
- Winter (bulan desember, januari, februari) banyak penerbangan yang cancelled, cuaca buruk sepertinya.
- Paling sedikit cancel Autumn (September - November)
- Paling banyak penerbangan di tengah Summer (6 **7** 8), karena cuaca sedang panas-panasnya, dan banyak yang liburan ke pantai

## No.7

In [12]:
# Masukan syntax jawaban di sini
airports[airports['AIRPORT'].str.lower().str.contains('john f. kennedy|northwest florida')]

IATA_CODE                                            AIRPORT         CITY  \
96        ECP    Northwest Florida Beaches International Airport  Panama City   
166       JFK  John F. Kennedy International Airport (New Yor...     New York   

    STATE COUNTRY  LATITUDE  LONGITUDE  
96     FL     USA       NaN        NaN  
166    NY     USA  40.63975  -73.77893

In [13]:
new_flights = flights.merge(airports, how = 'left', left_on='DESTINATION_AIRPORT', right_on='IATA_CODE')
jfk_ecp = new_flights[new_flights['ORIGIN_AIRPORT'].isin(['JFK','ECP'])]

In [14]:
soal_7 = jfk_ecp.groupby(['AIRPORT']).agg({'FLIGHT_NUMBER': ['count'],
                                           'CANCELLED': ['sum'],
                                           'DISTANCE': ['mean']
                                          }).reset_index()
soal_7.columns = ['DESTINATION_FROM_JFK_ECP', 'TOTAL_PENERBANGAN', 'TOTAL_CANCELLED', 'DISTANCE_MEAN']
soal_7.sort_values('TOTAL_PENERBANGAN', ascending=False)

DESTINATION_FROM_JFK_ECP  TOTAL_PENERBANGAN  \
38                San Francisco International Airport                 23   
19                  Los Angeles International Airport                 21   
11   Gen. Edward Lawrence Logan International Airport                 15   
27                      Orlando International Airport                  9   
4             Charlotte Douglas International Airport                  8   
24                        Miami International Airport                  8   
22                     McCarran International Airport                  7   
41                        Tampa International Airport                  7   
14   Hartsfield-Jackson Atlanta International Airport                  6   
28                   Palm Beach International Airport                  6   
10    Fort Lauderdale-Hollywood International Airport                  6   
3     Charleston International Airport/Charleston AFB                  5   
21             Luis Muñoz Marín International Airport                  4   
39               Seattle-Tacoma International Airport                  3   
40            Southwest Florida International Airport                  3   
1               Buffalo Niagara International Airport                  3   
29           Phoenix Sky Harbor International Airport                  3   
0              Austin-Bergstrom International Airport                  3   
6             Cleveland Hopkins International Airport                  3   
17                 Jacksonville International Airport                  3   
34          Ronald Reagan Washington National Airport                  2   
2                    Burlington International Airport                  2   
5                Chicago O'Hare International Airport                  2   
7                               Cyril E. King Airport                  2   
25                    Nashville International Airport                  2   
37  San Diego International Airport (Lindbergh Field)                  2   
9                        Denver International Airport                  2   
8                                   Dallas Love Field                  2   
35               Salt Lake City International Airport                  2   
32                           Rafael Hernández Airport                  2   
33               Raleigh-Durham International Airport                  2   
18               Long Beach Airport (Daugherty Field)                  1   
42            Washington Dulles International Airport                  1   
36                  San Antonio International Airport                  1   
31                     Portland International Airport                  1   
30                Port Columbus International Airport                  1   
12               George Bush Intercontinental Airport                  1   
26                      Oakland International Airport                  1   
13            Greater Rochester International Airport                  1   
23                                  Mercedita Airport                  1   
15                     Honolulu International Airport                  1   
16                 Indianapolis International Airport                  1   
20  Louis Armstrong New Orleans International Airport                  1   
43                           William P. Hobby Airport                  1   

    TOTAL_CANCELLED  DISTANCE_MEAN  
38                1           2586  
19                0           2475  
11                0            187  
27                0            944  
4                 1            541  
24                1           1089  
22                0           2248  
41                0           1005  
14                0            500  
28                0           1028  
10                0           1069  
3                 0            636  
21                0           1598  
39                0           2422  
40                1           1074  
1               

Destinasi airport mana yang paling banyak jumlah penerbangannya? **San Francisco International Airport**

# Soal Advanced

## No.8

In [15]:
# Masukan syntax jawaban di sini
tel_revenue = pd.read_csv('01 telecom_revenue.csv')
tel_usage = pd.read_csv('02 telecom_usage.csv')

In [16]:
print('Data revenue ada duplikat?', tel_revenue.duplicated(subset='CustomerID').any())
tel_revenue.head()

Data revenue ada duplikat? False


CustomerID Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
0     3164846    No           34.91           619.0                  18.0   
1     3203230   Yes           43.24           405.0                  40.0   
2     3371590    No           73.47          1281.0                  71.0   
3     3138302    No          776.27          3978.0                 160.0   
4     3375774    No          189.20          1474.0                  95.0   

   OverageMinutes  RoamingCalls Occupation  
0            53.0           0.0      Other  
1             0.0           0.0      Other  
2             0.0           0.0      Other  
3          2278.0          22.9      Other  
4             0.0           0.0      Other

In [17]:
print('Data usage ada duplikat?', tel_usage.duplicated(subset='CustomerID').any())
tel_usage.head()

Data usage ada duplikat? False


CustomerID  DroppedCalls  BlockedCalls  UnansweredCalls  CustomerCareCalls
0     3389874           0.0           1.0              7.7                0.7
1     3282042          15.3           1.7            152.0                5.7
2     3054454           8.7           0.3             30.3                0.0
3     3091842           6.0           1.7             19.3                0.0
4     3382382           0.7           0.7              5.0                0.0

In [18]:
monthly_revenue = tel_revenue['MonthlyRevenue'] > 10
occupation = tel_revenue['Occupation'].isin(['Professional', 'Student', 'Crafts'])
df1 = tel_revenue[monthly_revenue & occupation]

awal_genap = tel_usage['CustomerCareCalls'].astype(str).str.startswith(('2', '4', '6', '8'))
un_more_block = tel_usage['UnansweredCalls'] > tel_usage['BlockedCalls']
df2 = tel_usage[awal_genap & un_more_block]

In [19]:
soal_8 = df1.merge(df2, on='CustomerID')

In [20]:
soal_8_agg = soal_8.groupby('Occupation').agg({'CustomerID': 'nunique',
                                               'DroppedCalls': 'mean'
                                              }).reset_index()
soal_8_agg.columns = ['Occupation', 'total_CustomerID', 'mean_DroppedCalls']
soal_8_agg

Occupation  total_CustomerID  mean_DroppedCalls
0        Crafts                 4          10.600000
1  Professional                15           4.786667

pada occupation apa, rata-rata DroppedCalls paling besar? **Crafts**